src/app/credExplorer/weights/PluginWeightConfig.js

```diff
@@ -37,45 +37,43 @@ export class PluginWeightConfig extends React.Component<Props, State> {
   }
 
   _renderNodeWeightControls() {
-    return this.state.weights.nodes.map((wnt: WeightedNodeType) => {
-      const onChange = (newType: WeightedNodeType) => {
-        const index = this.state.weights.nodes.findIndex((x) =>
-          deepEqual(x.type, wnt.type)
+    return Array.from(this.state.weights.nodes.values()).map(
+      (wnt: WeightedNodeType) => {
+        const onChange = (newType: WeightedNodeType) => {
+          const newNodes = MapUtil.copy(this.state.weights.nodes);
```

Comment:
This computation needs to be in a `setState` callback, because you read
from `this.state` and use the result to construct the new state.

Something like:

```js
this.setState((state) => {
  const newNodes = MapUtil.copy(state.weights.nodes);
  newNodes.set(newType.type.prefix, newType);
  const weights = {...state.weights, newNodes};
  return {weights};
}, () => this.fire());
```

with various levels of inlining possible depending on your preference.

(Recall that React batches state updates, so the “pending state” may be
more recent than `this.state`. The code as is risks stomping updates.)

Likewise below for edges.

Reply:
Good catch. Done.